Building user-based recommendation model for Amazon.

Project 3

DESCRIPTION

The dataset provided contains movie reviews given by Amazon customers. Reviews were given between May 1996 and July 2014.

Data Dictionary
UserID – 4848 customers who provided a rating for each movie
Movie 1 to Movie 206 – 206 movies for which ratings are provided by 4848 distinct users

Data Considerations
- All the users have not watched all the movies and therefore, all movies are not rated. These missing values are represented by NA.
- Ratings are on a scale of -1 to 10 where -1 is the least rating and 10 is the best.

Analysis Task
- Exploratory Data Analysis:

    Which movies have maximum views/ratings?
    What is the average rating for each movie? Define the top 5 movies with the maximum ratings.
    Define the top 5 movies with the least audience.

- Recommendation Model: Some of the movies hadn’t been watched and therefore, are not rated by the users. Netflix would like to take this as an opportunity and build a machine learning recommendation algorithm which provides the ratings for each of the users.

    Divide the data into training and test data
    Build a recommendation model on training data
    Make predictions on the test data


In [395]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import sklearn

In [396]:
#Important to get multiple outputs in same cell window
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [397]:
df = pd.read_csv('J:/Simplii Learn DS/Machine Learning/Dataset for Project/Amazon - Movies and TV Ratings.csv')
df.head()

user_id  Movie1  Movie2  Movie3  Movie4  Movie5  Movie6  Movie7  \
0  A3R5OBKS7OM2IR     5.0     5.0     NaN     NaN     NaN     NaN     NaN   
1   AH3QC2PC1VTGP     NaN     NaN     2.0     NaN     NaN     NaN     NaN   
2  A3LKP6WPMP9UKX     NaN     NaN     NaN     5.0     NaN     NaN     NaN   
3   AVIY68KEPQ5ZD     NaN     NaN     NaN     5.0     NaN     NaN     NaN   
4  A1CV1WROP5KTTW     NaN     NaN     NaN     NaN     5.0     NaN     NaN   

   Movie8  Movie9  ...  Movie197  Movie198  Movie199  Movie200  Movie201  \
0     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
1     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
2     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
3     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
4     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   

   Movie202  Movie203  Movie204  Movie205  Movie206  
0       NaN       NaN       NaN       NaN       NaN  
1       NaN       NaN       NaN       NaN       NaN  
2       NaN       NaN       NaN       NaN       NaN  
3       NaN       NaN       NaN       NaN       NaN  
4       NaN       NaN       NaN       NaN       NaN  

[5 rows x 207 columns]

- Exploratory Data Analysis:

  *  Which movies have maximum views/ratings?
  *  What is the average rating for each movie? Define the top 5 movies with the maximum ratings.
  *  Define the top 5 movies with the least audience.


In [398]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4848 entries, 0 to 4847
Columns: 207 entries, user_id to Movie206
dtypes: float64(206), object(1)
memory usage: 7.7+ MB


In [399]:
df.describe()

Movie1  Movie2  Movie3  Movie4     Movie5  Movie6  Movie7  Movie8  \
count     1.0     1.0     1.0     2.0  29.000000     1.0     1.0     1.0   
mean      5.0     5.0     2.0     5.0   4.103448     4.0     5.0     5.0   
std       NaN     NaN     NaN     0.0   1.496301     NaN     NaN     NaN   
min       5.0     5.0     2.0     5.0   1.000000     4.0     5.0     5.0   
25%       5.0     5.0     2.0     5.0   4.000000     4.0     5.0     5.0   
50%       5.0     5.0     2.0     5.0   5.000000     4.0     5.0     5.0   
75%       5.0     5.0     2.0     5.0   5.000000     4.0     5.0     5.0   
max       5.0     5.0     2.0     5.0   5.000000     4.0     5.0     5.0   

       Movie9  Movie10  ...  Movie197  Movie198  Movie199  Movie200  Movie201  \
count     1.0      1.0  ...  5.000000       2.0       1.0  8.000000  3.000000   
mean      5.0      5.0  ...  3.800000       5.0       5.0  4.625000  4.333333   
std       NaN      NaN  ...  1.643168       0.0       NaN  0.517549  1.154701   
min       5.0      5.0  ...  1.000000       5.0       5.0  4.000000  3.000000   
25%       5.0      5.0  ...  4.000000       5.0       5.0  4.000000  4.000000   
50%       5.0      5.0  ...  4.000000       5.0       5.0  5.000000  5.000000   
75%       5.0      5.0  ...  5.000000       5.0       5.0  5.000000  5.000000   
max       5.0      5.0  ...  5.000000       5.0       5.0  5.000000  5.000000   

       Movie202  Movie203  Movie204   Movie205   Movie206  
count  6.000000       1.0  8.000000  35.000000  13.000000  
mean   4.333333       3.0  4.375000   4.628571   4.923077  
std    1.632993       NaN  1.407886   0.910259   0.277350  
min    1.000000       3.0  1.000000   1.000000   4.000000  
25%    5.000000       3.0  4.750000   5.000000   5.000000  
50%    5.000000       3.0  5.000000   5.000000   5.000000  
75%    5.000000       3.0  5.000000   5.000000   5.000000  
max    5.000000       3.0  5.000000   5.000000   5.000000  

[8 rows x 206 columns]

In [400]:
df.count()

user_id     4848
Movie1         1
Movie2         1
Movie3         1
Movie4         2
Movie5        29
Movie6         1
Movie7         1
Movie8         1
Movie9         1
Movie10        1
Movie11        2
Movie12        5
Movie13        1
Movie14        1
Movie15        1
Movie16      320
Movie17        1
Movie18        1
Movie19        2
Movie20        1
Movie21        1
Movie22        2
Movie23        3
Movie24        5
Movie25        1
Movie26        3
Movie27        1
Movie28        3
Movie29      243
            ... 
Movie177       1
Movie178       1
Movie179       7
Movie180       1
Movie181       2
Movie182      30
Movie183       1
Movie184      17
Movie185      24
Movie186       9
Movie187       1
Movie188       6
Movie189       5
Movie190       7
Movie191       6
Movie192      10
Movie193       7
Movie194       7
Movie195       1
Movie196       9
Movie197       5
Movie198       2
Movie199       1
Movie200       8
Movie201       3
Movie202       6
Movie203       1
Movie204      

In [401]:
df.count(axis=0).drop('user_id').max()

2313

In [402]:
df.count().drop('user_id').idxmax()

'Movie127'

Hence 'Movie127' has the highest views/ratings of 2313.

In [403]:
df.sum().drop('user_id')

Movie1         5
Movie2         5
Movie3         2
Movie4        10
Movie5       119
Movie6         4
Movie7         5
Movie8         5
Movie9         5
Movie10        5
Movie11       10
Movie12       25
Movie13        5
Movie14        4
Movie15        5
Movie16     1446
Movie17        3
Movie18        5
Movie19        7
Movie20        3
Movie21        5
Movie22       10
Movie23       15
Movie24       22
Movie25        5
Movie26        9
Movie27        5
Movie28       10
Movie29     1168
Movie30        9
            ... 
Movie177       5
Movie178       5
Movie179      33
Movie180       5
Movie181      10
Movie182     142
Movie183       5
Movie184      82
Movie185     115
Movie186      45
Movie187       5
Movie188      30
Movie189      25
Movie190      33
Movie191      30
Movie192      45
Movie193      32
Movie194      33
Movie195       4
Movie196      44
Movie197      19
Movie198      10
Movie199       5
Movie200      37
Movie201      13
Movie202      26
Movie203       3
Movie204      

In [404]:
df.columns

Index(['user_id', 'Movie1', 'Movie2', 'Movie3', 'Movie4', 'Movie5', 'Movie6',
       'Movie7', 'Movie8', 'Movie9',
       ...
       'Movie197', 'Movie198', 'Movie199', 'Movie200', 'Movie201', 'Movie202',
       'Movie203', 'Movie204', 'Movie205', 'Movie206'],
      dtype='object', length=207)

The average rating of all movies. 

In [405]:
df.mean()

Movie1      5.000000
Movie2      5.000000
Movie3      2.000000
Movie4      5.000000
Movie5      4.103448
Movie6      4.000000
Movie7      5.000000
Movie8      5.000000
Movie9      5.000000
Movie10     5.000000
Movie11     5.000000
Movie12     5.000000
Movie13     5.000000
Movie14     4.000000
Movie15     5.000000
Movie16     4.518750
Movie17     3.000000
Movie18     5.000000
Movie19     3.500000
Movie20     3.000000
Movie21     5.000000
Movie22     5.000000
Movie23     5.000000
Movie24     4.400000
Movie25     5.000000
Movie26     3.000000
Movie27     5.000000
Movie28     3.333333
Movie29     4.806584
Movie30     4.500000
              ...   
Movie177    5.000000
Movie178    5.000000
Movie179    4.714286
Movie180    5.000000
Movie181    5.000000
Movie182    4.733333
Movie183    5.000000
Movie184    4.823529
Movie185    4.791667
Movie186    5.000000
Movie187    5.000000
Movie188    5.000000
Movie189    5.000000
Movie190    4.714286
Movie191    5.000000
Movie192    4.500000
Movie193    4

the top 5 movies with the maximum ratings.

In [406]:
df1 = df.count().drop('user_id')
df1.nlargest()

Movie127    2313
Movie140     578
Movie16      320
Movie103     272
Movie29      243
dtype: int64

Define the top 5 movies with the least audience.

In [407]:
df1.nsmallest()

Movie1    1
Movie2    1
Movie3    1
Movie6    1
Movie7    1
dtype: int64

- Recommendation Model: Some of the movies hadn’t been watched and therefore, are not rated by the users. Netflix would like to take this as an opportunity and build a machine learning recommendation algorithm which provides the ratings for each of the users.

    Divide the data into training and test data
    Build a recommendation model on training data
    Make predictions on the test data


In [408]:
n_users

4848

In [409]:
n_movies

207

In [410]:
df.drop('user_id', axis = 1, inplace = True)
df['user_id'] = range(1,len(df)+1)
mid = df['user_id']
df.drop(labels=['user_id'], axis=1, inplace = True)
df.insert(0,'user_id',mid)
df.fillna(0, inplace = True)
df.head()

user_id  Movie1  Movie2  Movie3  Movie4  Movie5  Movie6  Movie7  Movie8  \
0        1     5.0     5.0     0.0     0.0     0.0     0.0     0.0     0.0   
1        2     0.0     0.0     2.0     0.0     0.0     0.0     0.0     0.0   
2        3     0.0     0.0     0.0     5.0     0.0     0.0     0.0     0.0   
3        4     0.0     0.0     0.0     5.0     0.0     0.0     0.0     0.0   
4        5     0.0     0.0     0.0     0.0     5.0     0.0     0.0     0.0   

   Movie9  ...  Movie197  Movie198  Movie199  Movie200  Movie201  Movie202  \
0     0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
1     0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
2     0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
3     0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
4     0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   

   Movie203  Movie204  Movie205  Movie206  
0       0.0       0.0       0.0       0.0  
1       0.0       0.0       0.0       0.0  
2       0.0       0.0       0.0       0.0  
3       0.0       0.0       0.0       0.0  
4       0.0       0.0       0.0       0.0  

[5 rows x 207 columns]

In [411]:
#Converting all columns to integer datatype.
df = df.astype(int)
print(df.dtypes)

user_id     int32
Movie1      int32
Movie2      int32
Movie3      int32
Movie4      int32
Movie5      int32
Movie6      int32
Movie7      int32
Movie8      int32
Movie9      int32
Movie10     int32
Movie11     int32
Movie12     int32
Movie13     int32
Movie14     int32
Movie15     int32
Movie16     int32
Movie17     int32
Movie18     int32
Movie19     int32
Movie20     int32
Movie21     int32
Movie22     int32
Movie23     int32
Movie24     int32
Movie25     int32
Movie26     int32
Movie27     int32
Movie28     int32
Movie29     int32
            ...  
Movie177    int32
Movie178    int32
Movie179    int32
Movie180    int32
Movie181    int32
Movie182    int32
Movie183    int32
Movie184    int32
Movie185    int32
Movie186    int32
Movie187    int32
Movie188    int32
Movie189    int32
Movie190    int32
Movie191    int32
Movie192    int32
Movie193    int32
Movie194    int32
Movie195    int32
Movie196    int32
Movie197    int32
Movie198    int32
Movie199    int32
Movie200    int32
Movie201  

In [412]:
from sklearn.model_selection import train_test_split
n_users = df['user_id'].count()
n_movies = df.columns.shape[0]
train_data,test_data = train_test_split(df, test_size = 0.25)

In [413]:
train_data_matrix = np.zeros((n_users, n_movies))
for line in train_data.itertuples():
    train_data_matrix[line[1]-1,line[2]-1] = line[3]
train_data_matrix

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [414]:
test_data_matrix = np.zeros((n_users, n_movies))
for line in train_data.itertuples():
    train_data_matrix[line[1]-1,line[2]-1] = line[3]
test_data_matrix

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [416]:
from sklearn.metrics import pairwise_distances
user_similarity = pairwise_distances(train_data_matrix, metric= 'cosine')
movie_similarity = pairwise_distances(train_data_matrix.T, metric = 'cosine')
mean_user_rating = train_data_matrix.mean(axis = 1)[:,np.newaxis]
ratings_diff = (train_data_matrix - mean_user_rating)
user_pred = mean_user_rating + user_similarity.dot(ratings_diff)/np.array([np.abs(user_similarity).
                                                                           sum(axis=1)]).T
user_pred

array([[ 2.41545894e-02,  2.41545894e-02,  2.41545894e-02, ...,
         2.41545894e-02,  2.41545894e-02,  2.41545894e-02],
       [-4.98341023e-06, -4.98341023e-06, -4.98341023e-06, ...,
        -4.98341023e-06, -4.98341023e-06, -4.98341023e-06],
       [-4.98341023e-06, -4.98341023e-06, -4.98341023e-06, ...,
        -4.98341023e-06, -4.98341023e-06, -4.98341023e-06],
       ...,
       [-4.98341023e-06, -4.98341023e-06, -4.98341023e-06, ...,
        -4.98341023e-06, -4.98341023e-06, -4.98341023e-06],
       [-4.98341023e-06, -4.98341023e-06, -4.98341023e-06, ...,
        -4.98341023e-06, -4.98341023e-06, -4.98341023e-06],
       [-4.98341023e-06, -4.98341023e-06, -4.98341023e-06, ...,
        -4.98341023e-06, -4.98341023e-06, -4.98341023e-06]])